nip47/cipher/cipher.go

```diff
@@ -0,0 +1,112 @@
+package cipher
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/getAlby/hub/nip47/models"
+	"github.com/nbd-wtf/go-nostr/nip04"
+	"github.com/nbd-wtf/go-nostr/nip44"
+)
+
+type Nip47Cipher struct {
+	Version         string
+	pubkey          string
+	privkey         string
+	sharedSecret    []byte
+	conversationKey [32]byte
+}
+
+func NewNip47Cipher(version, pubkey, privkey string) (*Nip47Cipher, error) {
+	_, err := isVersionSupported(version)
+	if err != nil {
+		return nil, err
+	}
+
+	var ss []byte
+	var ck [32]byte
+	if version == "0.0" {
+		ss, err = nip04.ComputeSharedSecret(pubkey, privkey)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		ck, err = nip44.GenerateConversationKey(pubkey, privkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &Nip47Cipher{
+		Version:         version,
+		pubkey:          pubkey,
+		privkey:         privkey,
+		sharedSecret:    ss,
+		conversationKey: ck,
+	}, nil
+}
+
+func (c *Nip47Cipher) Encrypt(message string) (msg string, err error) {
+	if c.Version == "0.0" {
+		msg, err = nip04.Encrypt(message, c.sharedSecret)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		msg, err = nip44.Encrypt(message, c.conversationKey)
+		if err != nil {
+			return "", err
+		}
+	}
+	return msg, nil
+}
+
+func (c *Nip47Cipher) Decrypt(content string) (payload string, err error) {
+	if c.Version == "0.0" {
+		payload, err = nip04.Decrypt(content, c.sharedSecret)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		payload, err = nip44.Decrypt(content, c.conversationKey)
+		if err != nil {
+			return "", err
+		}
+	}
+	return payload, nil
+}
+
+func isVersionSupported(version string) (bool, error) {
+	versionParts := strings.Split(version, ".")
```

Comment:
I chatted with @bumi and the versioning is a real exception (and hopefully all apps update to nip44 so we can remove the v0.0 code)

So we think for now it's better just to check `version == "1.0" || version == "0.0"`

Reply:
Done!